README
           FSPv2 -> HTTP/1.1 protocol proxy
                     Radim Kolar
                       2003-9
        http://fsp.sourceforge.net/fsproxy.html
          This software is part of FSP project

Copyright:
You may copy or modify this file in any manner you wish, provided
that this notice is always included, and that you hold the author
harmless for any loss or damage resulting from the installation or
use of this software.

Summary:
This proxy supports HTTP/1.1 on output side and FSP v2 on input side.
FSP v2 over HTTP/1.1. This proxy do not caches anything.

In detail:
Using this proxy, you can browse FSP sites by any web browser
which support gopher protocol.
This proxy DOES NOT tunnels HTTP over FSP. This is a different project.

Installation:

1. Get FSP protocol stack for Java - fsp javalib

This program needs correctly installed (in CLASSPATH) jfsplib,
you can get it from main FSP protocol suite site:
http://fsp.sourceforge.net/javalib.html
You need at least 1.0rc7 version of jfsplib

2. Setup WWW browser's proxy
Setup Gopher proxy in web browser to port 9090 on host which is running
fsproxy.

3. Run fsp proxy daemon

distributed fsproxy.jar is autostartable, but there seems to be problem
that java launcher (tested on 1.6.0_16) is unable to merge classpath with
autostartable jar content.

Running:
Usage: java -cp fsproxy-0.X.jar;jfsplib-1.0rc8.jar fspproxy [portnumber] [mime type table]

Note: Unix systems are using : instead of ; as classpath separator

FSP proxy can load mime.types style file for mapping file extensions
to mime types. You can get this file from apache httpd distribution.

Examples:
Unix
   java -cp jfsplib-1.0rc8.jar:fsproxy-0.8.jar 9090 /usr/local/etc/mime.types

Windows
   C:\target>java -cp jfsplib-1.0rc8.jar;fsproxy-0.8.jar fspproxy
   Fri Oct 30 13:48:08 CET 2009 FSP Proxy 0.8 ready on 9090/*

Operation:

FSP Proxy runs on port by default 9090 and accepts fsp://*, http://*, ftp://*
and gopher://* URLs on input side. These requests are transformed info FSP v2
requests and sent to the target FSP server.  Most browser do not recognize
fsp:// urls, so you have to type gopher: instead of fsp:

Simply type in your browser URL bar 'gopher://your.favourite.fsp.site/' without
quotes and begin to browse content of particular FSP site.
In most cases you must use port number as part of URL:
   gopher://your.favourite.fsp.site:2000/

Directory listing is pretty fast, isn't it?

     **********************************************
         This is a free software.  Be creative.
        Let me know of any bugs and suggestions.
               *************************

Radim Kolar
hsn@users.sourceforge.net
Main developer of FSP project

CHANGES
0.9
    Use jfsplib from Maven central
    pom.xml plugins updated
    Prepare pom.xml for deployment to Maven Central
    Build fat jar for .zip file distribution
    Rename README to README.txt
    add CHANGES to .zip distribution
    Updated documentation in README
0.8
    Source files moved to src/ subdirectory
    Maven 2.2 build system added
    GNU makefile and SCons build systems removed
    TODO file removed, issues moved to mantis bug tracker
    Distributed jar file is autostartable
    return error to client if Range: header parsing failed
    detect on startup if jfsplib 1.0rc7 or greater is installed
0.7
   Nice directory listings with file size and last modification time!
        Added tagline to generated error messages
   FSP timeout increased from 30 to 500 sec.
   bittorent mime type added
   ftp:// url is accepted as alias
   send 502 not 401 on failed directory listing because we
     do not supports FSP passwords.
0.6
   Added link to FSP homepage into directory listings tagline
   Directory listings are now very fast. There are just
     1 RTT far away!
   Added help wanted note to every directory listings.
   You need to get 1.0rc5 Java FSP support library
   Added HTTP/1.1 protocol support!
   HTTP/1.1 Range requests are supported!
   Download managers in browsers now works very nicely with
     all that resuming stuff.
   Added note about help wanted with adding FSP to lftp,mozilla,wget..
0.5
   updated copyrights
   print usage on error
   more resistent to errors in build environment
   can load mime types from user-defined file
0.4
   directory should ends with / - send redirect if needed
   parent directory link in dirlisting fixed
   support for guessing Content-types from filenames
   support for Last-Modified date
0.3
   first public release - 'technology demo'


src/fspreq.java
/*
Copyright (c) 2003-2004 by Radim HSN Kolar (hsn@netmag.cz)

You may copy or modify this file in any manner you wish, provided
that this notice is always included, and that you hold the author
harmless for any loss or damage resulting from the installation or
use of this software.

		     This is a free software.  Be creative.
		    Let me know of any bugs and suggestions.
*/
import java.net.*;
import java.io.*;
import java.util.StringTokenizer;
import java.util.Date;
import java.util.Locale;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import net.sf.fsp.fsplib.FSPsession;
import net.sf.fsp.fsplib.FSPstat;
import net.sf.fsp.fsplib.FSPutil;

public class fspreq implements Runnable
{
    private Socket s;
    private boolean http10;
    private DataInputStream in;
    private DataOutputStream ou;
    private long rangestart=0;     /* where to start download */
    private long rangelength=-1;   /* how many bytes to download */
    private long rangefilesize=-1; /* expected filesize */

    private static DateFormat formatter= new SimpleDateFormat("dd-MMM-yyyy hh:mm", Locale.US);
    public final static int NAMELEN=30;
    public final static int DATELEN=20;

    fspreq(Socket ns)
    {
	s=ns;
    }

    public void run()
    {
	http10=true;

	try
	{
	    in=new DataInputStream (new BufferedInputStream(s.getInputStream()));
	    ou=new DataOutputStream(new BufferedOutputStream(s.getOutputStream(),4096));

	/* precteme radku GET / */

	s.setSoTimeout(fspproxy.client_timeout);

	String req=in.readLine();
	if(req==null)
	{
	    s.close();ou.close();s=null;ou=null;
	    return;
	}

        if(fspproxy.trace_url==true)
        {
            System.out.println("[FSPPROXY TRACE "+Thread.currentThread().getName()+"] > "+req);
        }

        int req_method;
	long ims=0;
	req_method=req.indexOf(" HTTP/1.",0);
	if (req_method==-1) http10=false;
	else
          while(true)
          {
	      String line;
	      /* read rest of HTTP headers */
	      line=in.readLine();
              if(line==null) break;
              if(line.length()==0) break;
	      String s1,s2;
	      int j;
	      j=line.indexOf(':',0);
	      if(j==-1) continue;
	      s1=line.substring(0,j).toLowerCase();
	      s2=line.substring(j+1);

              if(s1.equals("if-modified-since"))
	      {
                /* cut of file size */
               	j=s2.indexOf(';',0);
		if(j>-1) s2=s2.substring(0,j);
		try
		{
		   ims=new Date(s2).getTime();
	        }
	        catch (IllegalArgumentException baddate)                                        {}
		continue;
	      }
              if(s1.equals("range"))
	      {
                  /* Range: bytes = 2 - 5 / 23 */
		  try
		  {
		      String r;
		      long l;
		      int i;
		      byte state;
		      s2=s2.toLowerCase();
		      i=s2.indexOf("bytes");
		      if(i==-1)
			  throw new Exception();
		      for(i=i+5;;i++)
			  if ( s2.charAt(i)=='=')
			  {
			      s2=s2.substring(i+1).trim();
			      break;
			  }

		      //System.out.println("parse start "+s2);
		      if(s2.charAt(0)=='-')
	              {
			  rangestart = -1;
			  state = 1;
		      }
		      else
			  state = 0;
		      StringTokenizer st=new StringTokenizer(s2," \t\r\n-/");
		      while(st.hasMoreTokens())
		      {
		          r=st.nextToken();
		          //System.out.println("r2 "+r);
                          l=Long.valueOf(r).longValue();
			  switch(state)
			  {
			       case 0:
			           rangestart=l;
			           break;
			       case 1:
			           if ( rangestart == -1 )
				       rangelength = l;
				   else
			               rangelength=l-rangestart+1;
			           break;
			       case 2:
			           rangefilesize=l;
				   break;
			       default:
			           throw new Exception();
			   }
			   state++;
		      }
                  }
		  catch (Exception e)
		  {
		    rangestart=0;
		    rangelength=-1;
		    rangefilesize=-1;
	     	fspproxy.send_error(http10?10:9,416,"Unable to parse Range header in request.",ou);
		  }
	      }

          }
	String req2=null;

	if(!req.startsWith("GET "))
	{
	     fspproxy.send_error(http10?10:9,501,"This FSP proxy server supports only GET access method.",ou);
	}
	/* access check
	if(!mgr.checkInetAdr(s.getInetAddress().getAddress())) {
	      httpreq.server_error(http10?10:9,403,"Cache access denied.",ou);
	   }
	*/

	/* extract URL as req2 */
	int space;
	space=req.indexOf(' ');
        if(space==-1)
	    fspproxy.send_error(http10?10:9,400,"Can not find method and URI in request",ou);
	if(req_method==-1)
	    req2=req.substring(space+1);
	else
	    req2=req.substring(space+1,req_method);

        req2=req2.trim();
	if(req2.length()==0)
	    fspproxy.send_error(http10?10:9,400,"Can not find URI in request",ou);

	if(req2.charAt(0)=='/')
	{
	    fspproxy.send_error(http10?10:9,400,"Only proxy requests are supported.",ou);
	}

	/* parse request */
	String parsed[]=fspproxy.parseURL(req2,null);
	if(parsed[4]==null ||
		(!parsed[4].equalsIgnoreCase("fsp") &&
		 !parsed[4].equalsIgnoreCase("gopher") &&
		 !parsed[4].equalsIgnoreCase("ftp") &&
		 !parsed[4].equalsIgnoreCase("http")
		)
	  )
	{
	    fspproxy.send_error(http10?10:9,501,"This proxy speaks only FSP v2 protocol. A fsp://, gopher://, ftp://, http:// URLs are accepted on input.",ou);
	}

	/* create FSP session to host */
	FSPsession ses;
	FSPstat stat;
	if(parsed[1]==null) parsed[1]="0";
	try
	{
	    ses=new FSPsession(parsed[0],Integer.valueOf(parsed[1]).intValue());
	}
	catch (Exception en)
	{
	    fspproxy.send_error(http10?10:9,500,"Can not open session to FSP site "+parsed[0]+":"+parsed[1]+" Reason: "+en,ou);
	    return;
	}

        /*
	System.out.println("p2="+parsed[2]);
	System.out.println("p3="+parsed[3]);
	*/

	/* try to avoid stat */
	if(parsed[2].endsWith("/") && parsed[3].length()==0)
	{
	    /* fake stat info */
	    stat=new FSPstat();
	    stat.name=parsed[2];
	    stat.type=FSPstat.RDTYPE_DIR;
	}
	else
	{
	    /* stat the URL */
	    try
	    {
		stat=FSPutil.stat(ses,parsed[2]+parsed[3]);
	    }
	    catch (IOException io)
	    {
		fspproxy.send_error(http10?10:9,500,"Error getting file info. Reason: "+io,ou);
		return;
	    }
	}

	if(stat==null)
	{
	    /* file not found! */
	    fspproxy.send_error(http10?10:9,404,"File not found",ou);
	}
	/* Check IMS */
	if(ims>0 && stat.lastmod<=ims)
	{
	    /* not modified! */
	    ou.writeBytes("HTTP/1.0 304 YahOO\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\n\r\n");
	    ou.close();
	    return;
	}
	/* handle file request */
	if(stat.type==FSPstat.RDTYPE_DIR)
	{
	    if(!stat.name.endsWith("/"))
	    {
		/* send redirect */
		ou.writeBytes("HTTP/1.0 301 Moved\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\nLocation: "+stat.name+"/\r\n\r\n");
		ou.close();
		return;
	    }

	    /* generate a directory listing */
	    FSPstat list[]=null;

	    try {
		    list=FSPutil.statlist(ses,stat.name);
	    }
	    catch (IOException e) {
		       fspproxy.send_error(http10?10:9,502,"Can't get directory listing of "+stat.name,ou);
	    }

	    ou.writeBytes("HTTP/1.1 200 Listing\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\nContent-Type: text/html\r\nConnection: Close\r\nAccept-Ranges: bytes\r\n\r\n");
	    /* write title */
	    ou.writeBytes("<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 3.2 Final//EN\">\n<HTML>\n<HEAD>\n<TITLE>Index of ");
	    ou.writeBytes(stat.name);
	    ou.writeBytes("</TITLE>\n</HEAD>\n<BODY>\n<H1>Index of ");
	    ou.writeBytes(stat.name);
	    ou.writeBytes("</H1>\n<PRE>      ");
            ou.writeBytes(text_align("Name",NAMELEN));
	    ou.writeBytes(" ");
	    ou.writeBytes(text_align("Last modified",DATELEN));
	    ou.writeBytes("Size\n<HR>\n");
	    for(int i=0;i<list.length;i++)
	    {
		if(list[i].name.equals(".")) continue;
		if(list[i].name.equals(".."))
		{
		    if(stat.name.equals("/"))
			list[i].name="";
		    else
			list[i].name=stat.name.substring(0,stat.name.lastIndexOf('/',stat.name.length()-2));
		    ou.writeBytes("[DIR] <a href=\""+list[i].name+"/\">Parent directory</a>\n");
		    continue;
		    /*
		    ou.writeBytes("[parent]");
		    */
		}
		if(list[i].type==FSPstat.RDTYPE_DIR)
		{
		    ou.writeBytes("[DIR] ");
		    list[i].name+="/";
		} else
		    ou.writeBytes("[TXT] ");
		int nlen;
		nlen=list[i].name.length();
		ou.writeBytes("<a href=\""+list[i].name+"\">"+list[i].name.substring(0,Math.min(NAMELEN,nlen))+"</a> ");
		for(int q=NAMELEN-nlen;q>0;q--)
		    ou.write(' ');

                ou.writeBytes(text_align(formatter.format(new Date(list[i].lastmod)),DATELEN));
		ou.writeBytes(""+list[i].length/1024);
		ou.writeBytes("k\n");
	    }
	    ou.writeBytes("</PRE>\n<HR>\n");

	    ou.writeBytes("Listing generated by <a href=\"http://fsp.sourceforge.net/fsproxy.html\">"+fsploop.NAME+" "+fsploop.VERSION+"</a>.\n");
	    ou.close();
	    return;
	}

	/* transfer file */

	/* do we want range transfer? */
	if(rangestart>0 || rangelength > -1)
	{
	    /* yes, check user suplied valuez */
	    if ( rangestart == -1 )
		rangestart = stat.length - rangelength;
	    if (rangelength == -1 )
		rangelength = stat.length - rangestart;
	    if (
		rangestart + rangelength > stat.length ||
		(rangefilesize != stat.length && rangefilesize>=0 )
	       )
	     	    fspproxy.send_error(http10?10:9,416,"Specified Range is incorrect for this file.",ou);
            /* Do partial file transfer */
	    ou.writeBytes("HTTP/1.1 206 Partial\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\nAccept-Ranges: bytes\r\nContent-Type: "+fspproxy.guessContentType(stat.name)+"\r\nContent-length: "+rangelength+"\r\nContent-Range: bytes "+rangestart+"-"+(rangestart+rangelength-1)+"/"+stat.length+"\r\nConnection: Close\r\nLast-Modified: "+new Date(stat.lastmod).toGMTString()+"\r\n\r\n");
	    FSPutil.download(ses,stat.name,ou,rangestart,rangelength);
	    ou.close();
	    return;
	}

	ou.writeBytes("HTTP/1.1 200 Transfering\r\nConnection: Close\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\nAccept-Ranges: bytes\r\nContent-Type: "+fspproxy.guessContentType(stat.name)+"\r\nContent-length: "+stat.length+"\r\nLast-Modified: "+new Date(stat.lastmod).toGMTString()+"\r\n\r\n");
	FSPutil.download(ses,stat.name,ou,0,-1);
	ou.close();
	return;
    }
    catch (IOException err)
    {
	try
	{
	    // linger close
	    //System.out.println(err);
	    s.setSoLinger(true,0);
	    s.close();
	}
	catch (IOException ignore) {}
    }
}

final static public String text_align(String text,int data)
{
    if(text.length()>=data) return text.substring(0,data);
    StringBuffer sb=new StringBuffer(data);
    sb.append(text);
    for(int i=data-text.length();i>0;i--)
	sb.append(' ');
    return sb.toString();
}

}


src/fsploop.java
/*
Copyright (c) 2003-2004 by Radim HSN Kolar (hsn@netmag.cz)

You may copy or modify this file in any manner you wish, provided
that this notice is always included, and that you hold the author
harmless for any loss or damage resulting from the installation or
use of this software.

		     This is a free software.  Be creative.
		    Let me know of any bugs and suggestions.
*/
import java.net.*;
import java.io.IOException;
import java.util.Date;

public class fsploop implements Runnable
{
    public static final String  VERSION="0.9";
    public static final String  NAME="FSP Proxy";

    public static int MAXCLIENTS=15;

    /* port for running proxy server */
    public int serverport;
    public InetAddress serveradr;

    private ServerSocket server;

    fsploop(int port, InetAddress adr)
    {
	serverport=port;
	serveradr=adr;

	if(serverport==0) return;
	/* create server socket */
	if(serveradr!=null)
	    try {
		 server=new ServerSocket(serverport,10,serveradr);
	    }
	    catch(IOException e)
	    {
		  System.err.println("[FSPPROXY] ERROR: Cannot bind to port "+serverport+"/"+serveradr.getHostAddress()+"  Reason: "+e);
		  return;
	    }
	 else
	     try{
		     server=new ServerSocket(serverport);
		 }
	     catch(IOException e)
	     {
		       System.err.println("[FSPPROXY] ERROR: Cannot bind to my port "+serverport+"/*  Reason: "+e);
		       return;
	     }
    }

    public final void run()
    {
      ThreadGroup clients;
      fspreq client;
      Socket clientSocket;

      if(server==null) return;

      clients=new ThreadGroup("FSPPROXY-http-clients");
      System.out.println(new Date()+" "+NAME+" "+VERSION+" ready on "+serverport+"/"+(serveradr==null?"*":serveradr.getHostAddress()));
       while (true)
       {
		clientSocket = null;
		client=null;
		try
		{
		    clientSocket = server.accept();
		}
		catch (IOException e)
		{
		    System.err.println("Warning: Accept failed: "+e);
		    continue;
		}
		if(clients.activeCount()>MAXCLIENTS)
		  try
		  {
		      System.err.println(new Date()+" [FSPPROXY] Warning: Active connections limit ("+MAXCLIENTS+") reached, HTTP request rejected.");
		      clientSocket.setSoLinger(true,0);
		      clientSocket.close();
		  }
		  catch(IOException e) {}
		  finally { continue;}
		client=new fspreq(clientSocket);
		new Thread(clients,client).start();
       }  /* listen */
    }
}


src/fspproxy.java
/*
Copyright (c) 2003-2004 by Radim HSN Kolar (hsn@netmag.cz)

You may copy or modify this file in any manner you wish, provided
that this notice is always included, and that you hold the author
harmless for any loss or damage resulting from the installation or
use of this software.

		     This is a free software.  Be creative.
		    Let me know of any bugs and suggestions.
*/
import java.io.*;
import java.util.StringTokenizer;
import java.util.Date;

/* stand-alone version of fsp -> http proxy server */
public class fspproxy
{

    public static int client_timeout;
    public static boolean trace_url;
    public static String defaultype="application/octet-stream";

    public static void print_usage()
    {
	System.out.println("java fspproxy [port number] [mime types file]");
	System.exit(1);
    }

    public static void main(String argv[])
    {
    /* check if jfsplib is installed */
    try
    {
    	Class.forName("net.sf.fsp.fsplib.FSPprotection");
    }
    catch (ClassNotFoundException z)
    {
    	System.out.println("jfsplib was not found in your classpath or it is too old version.\nMinimum required version for FSP Proxy is jfsplib 1.0rc7.\nSee http://fsp.sourceforge.net/javalib.html for more info.");
    	System.exit(2);
    }
	client_timeout=500000;
	trace_url=true;
	int port=9090;
	if(argv.length>0)
	{
	    try
	    {
	       port=Integer.valueOf(argv[0]).intValue();
	    }
	    catch (Exception z)
	    {
		print_usage();
            }
	    if(argv.length>1)
	    {
                System.out.println(new Date()+" Using user defined MIME types table from "+argv[1]);
                loadMimeTypes(argv[1]);
	    }
	}
	new fsploop(port,null).run();
    }

    /** generates error message and closes the connection */
    public final static void send_error(int httpversion,int rc,String err,
	   DataOutputStream ou) throws IOException
    {
       StringBuffer bd=new StringBuffer(1024); // body

       bd.append("<!DOCTYPE HTML PUBLIC \"-//IETF//DTD HTML 2.0//EN\">\n<HTML><HEAD><TITLE>Error</TITLE></HEAD>\n<BODY><h2>");
       bd.append(rc);
       bd.append(" ");
       bd.append(err);
       bd.append("</h2>\r\n");
       bd.append("Generated by <a href=\"http://fsp.sourceforge.net/fsproxy.html\">"+fsploop.NAME+" "+fsploop.VERSION+"</a>.\n");
       bd.append("</BODY></HTML>\r\n");

       if(httpversion>9)
       {
	   /* write HTTP header */
	   ou.writeBytes("HTTP/1.0 "+rc+" FSPPROXY\r\nContent-Type: text/html\r\nServer: "+fsploop.NAME+" "+fsploop.VERSION+"\r\n\r\n");
       }
       ou.writeBytes(bd.toString());
       ou.close();

       throw new EOFException("Request processing canceled");
    }

    /*  prechrousta URL a vrati
    0 - hostname
    1 - port (if any) jinak null
    2 - directory
    3 - file including query string ("" if empty)
    4 - protocol (null if http)
    */

    /* code from Smart Cache mgr.java */
    final public static String[] parseURL(String url,String proto)
    {
      String[] res=new String[5];
      res[3]=""; /* HashTable do not likes NULL */
      int i,j,seven;

      if(proto!=null)
      {
      /* we allready have protocol parsed */
      res[4]=proto;
      seven=proto.length()+3; /* '://' */

      }
      else
      {
	  seven=url.indexOf("://");
	  res[4]=url.substring(0,seven);
	  seven+=3;
      }
      /* http -> null */
      if(seven==7)
       {
	char c;
	c=res[4].charAt(0);
	if(c=='h' || c=='H') res[4]=null;
       }

      i=url.indexOf('/',seven);
      if(i==-1) { url+='/'; i=url.length()-1;}

      j=url.indexOf(':',seven);
      /* http://zero.dsd:434/ */
      if(j!=-1 && j<i)  /* mame tu portname */
			{
			  res[0]=url.substring(seven,j).toLowerCase();
			  if(j+1<i)
			  {
			    res[1]=url.substring(j+1,i);
			    if(res[1].equals("80")) res[1]=null;
			  }
			}
	       else
		 {
		  res[0]=url.substring(seven,i).toLowerCase();
		 }

      /* parse adr */

      /* najdeme nejvice vlevo z moznych problemovych znaku */

       j=url.length()-1;
       byte v[];
       v=new byte[j+1];
       url.getBytes(i,j+1,v,i);
       // getChars(i,j,v,0);
       loop1:for(int zz=i;zz<=j;zz++)
	{
	switch(v[zz])
	{
	case 0x3b: // ;
	case 0x3a: // :
	case 0x3d: // =
	case 0x3f: // ?
		  j=zz;break loop1;
	case 0x23: // #
	 url=url.substring(0,zz);break loop1;
	}
	}


       j=url.lastIndexOf('/',j);

      String adr=url.substring(i,j+1); // adresar
      /* Normalize URL - remove /../ */
      /* NOTE: /./ should be also removed, but it do not hurt anything? */
      while( (i=adr.indexOf("/../")) >= 0 )
      {
       int l;
       if( (l=adr.lastIndexOf('/',i-1)) >=0 )
	  adr= adr.substring(0,l)+ adr.substring(i+3);
       else
	 adr=adr.substring(i+3);
      }
      res[2]=adr;
      if(j+1!=url.length()) res[3]=url.substring(j+1);
      return res;
    }

 public final static void loadMimeTypes(String fname)
 {
  if(fname==null) return;
  File f=new File(fname);
  if(!f.isFile())
  {
      System.err.println("[ERROR] Can't load mime.types from file '"+fname+"'");
      return;
  }
  try
  {
    BufferedReader in=new BufferedReader(new LineNumberReader(new FileReader(fname)));
    String line;
    StringTokenizer st;
    String mimetype,ext;
    // init GT
    guesstable=new String[0];
    while(true)
    {
	    ext=mimetype=null;
	    line=in.readLine();
	    if(line==null) break;
	    st=new StringTokenizer(line);
	    if(!st.hasMoreTokens()) continue;
	    mimetype=st.nextToken();
	    if(mimetype.startsWith("#")) continue;
	    while(true)
	    {

	    	if(!st.hasMoreTokens()) break;
	    	ext=st.nextToken();
		updateGuessTable(mimetype,ext);
	    }
    }
    in.close();
  }
  catch (IOException grrrrrrrrrrrrr)
   {
    System.err.println("[ERROR] Reading mime.types from "+fname);
   }
 }

 public final static String guessContentType(String fname)
 {
  fname=fname.toLowerCase();
  for(int i=0;i<guesstable.length;i+=2)
   {
    if(fname.endsWith(guesstable[i])) return guesstable[i+1];
   }
  // System.out.println("[WARNING] Can not determine MIME type for "+fname+", defaulting to "+defaultype);
  return defaultype;
 }

 /* pozdejsi koncovka prepise predchozi hodnotu */
 private final static void updateGuessTable(String mimetype,String ext)
 {
 	 if(mimetype==null || ext==null) return;
	 if(mimetype.length()==0 || ext.length()==0) return;
	 ext=("."+ext).toLowerCase();
	 for(int i=0;i<guesstable.length;i+=2)
	 {
		 if(ext.equals(guesstable[i]))
		   {
		     guesstable[i+1]=mimetype;
		     return;
		   }
	 }
	 String tmp[];
	 tmp=new String[guesstable.length+2];
	 System.arraycopy(guesstable,0,tmp,0,guesstable.length);
	 tmp[guesstable.length]=ext;
	 tmp[guesstable.length+1]=mimetype;
	 guesstable=tmp;
}

 static String guesstable[]=
           {
             // java stuff
             // ".java","application/java",
             ".java","text/plain",
             ".class","application/java-vm",
             ".jar","application/java-archive",

	     // images
             ".gif","image/gif",
	     ".ief","image/ief",
	     ".tiff","image/tiff",
	     ".tif","image/tiff",
             ".jpeg","image/jpeg",
             ".jpe","image/jpeg",
             ".jpg","image/jpeg",
	     ".png","image/png",
	     ".ras","image/x-cmu-raster",
	     ".bmp","image/x-ms-bmp",
	     ".pnm","image/x-portable-anymap",
	     ".pbm","image/x-portable-bitmap",
	     ".pgm","image/x-portable-graymap",
	     ".ppm","image/x-portable-pixmap",
	     ".rgb","image/x-rgb",
	     ".xbm","image/x-xbitmap",
	     ".xpm","image/x-xpixmap",
	     ".xwd","image/x-xwindowdump",

	     // plain text
             ".txt","text/plain",
             ".text","text/plain",
             ".doc","text/plain",
             ".log","text/plain",
	     ".csv","text/comma-separated-values",
	     ".tsv","text/tab-separated-values",

	     // hypertext
             "welcome",  "text/html", // cache-generated index
	     ".css","text/css",
	     ".js","text/javascript",
	     ".pl",  "text/html",
	     ".cgi",  "text/html",
	     ".asp",  "text/html",
	     ".jsp",  "text/html",
             ".htm",  "text/html",
             ".html",  "text/html",
	     ".shtml", "text/html",
             ".htmli",  "text/html",
             ".dchtml",  "text/html",
	     ".pht",   "text/html",
	     ".phtml", "text/html",
	     ".php",   "text/html",
	     ".php3",  "text/html",
	     ".php4",  "text/html",
	     ".php3p", "text/html",
	     ".php4p", "text/html",

	     ".texi","application/x-texinfo",
	     ".texinfo","application/x-texinfo",

	     // VRML
	     ".vrm","x-world/x-vrml",
	     ".vrml","x-world/x-vrml",
	     ".wrl" ,"x-world/x-vrml",

	     // formated text
	     ".rtx","text/richtext",
             ".pdf","application/pdf",
	     ".rtf","application/rtf",
	     ".ai","application/postscript",
	     ".ps","application/postscript",
	     ".eps","application/postscript",
	     ".wp5","application/wordperfect5.1",
	     ".wk","application/x-123",
	     ".dvi","application/x-dvi",
	     ".frm","application/x-maker",
	     ".maker","application/x-maker",
	     ".frame","application/x-maker",
	     ".fm"   ,"application/x-maker",
	     ".fb"   ,"application/x-maker",
	     ".book", "application/x-maker",
	     ".fbdoc","application/x-maker",

	     // fonts
	     ".pfa","application/x-font",
	     ".pfb","application/x-font",
	     ".gsf","application/x-font",
	     ".pcf","application/x-font",
	     ".pcf.z","application/x-font",
	     ".gf","application/x-tex-gf",
	     ".pk","application/x-tex-pk",

	     // archives
             ".zip","application/zip",
             ".tar","application/x-tar",
	     ".hqx","application/mac-binhex40",
	     ".bcpio","application/x-bcpio",
	     ".cpio","application/x-cpio",
	     ".deb","application/x-debian-package",
             ".gtar","application/x-gtar",
	     ".tgz" ,"application/x-gtar",
	     ".tar.gz","application/x-gtar",
	     ".shar","application/x-shar",
	     ".sit","application/x-stuffit",
	     ".sv4cpio","application/x-sv4cpio",
	     ".sv4crc","application/x-sv4crc",

	     //audio
	     ".au","audio/basic",
	     ".snd","audio/basic",
             ".mid","audio/midi",
             ".midi","audio/midi",
             ".mp2","audio/mpeg",
             ".mpega","audio/mpeg",
             ".mpga","audio/mpeg",
             ".mp3","audio/mpeg",
	     ".m3u","audio/mpegurl",
             ".aif","audio/x-aiff",
             ".aiff","audio/x-aiff",
             ".aifc","audio/x-aiff",
	     ".wav","audio/x-wav",
	     ".ra" ,"audio/x-pn-realaudio",
	     ".rm" ,"audio/x-pn-realaudio",
	     ".ram","audio/x-pn-realaudio",

	     //video
	     ".mpeg","video/mpeg",
	     ".mpg","video/mpeg",
             ".mpe","video/mpeg",
             ".qt","video/quicktime",
             ".mov","video/quicktime",
             ".avi","video/x-msvideo",
             ".movie","video/x-sgi-movie",
	     ".dl","video/dl",
	     ".fli","video/fli",
	     ".gl","video/gl",
	     ".asf","video/x-ms-asf",
	     ".asx","video/x-ms-asf",

	     //sources
	     ".tex","text/x-tex",
	     ".c","text/plain",
	     ".cc","text/plain",
	     ".cpp","text/plain",
	     ".h","text/plain",
	     ".hpp","text/plain",
	     ".py","text/plain",
	     ".ltx","text/x-tex",
	     ".sty","text/x-tex",
	     ".cls","text/x-tex",
	     ".latex","application/x-latex",
	     ".oda","application/oda",
	     ".t"  ,"application/x-troff",
	     ".tr" ,"application/x-troff",
	     ".roff","application/x-troff",
	     ".man", "application/x-troff-man",
	     ".me",  "application/x-troff-me",
	     ".ms",  "application/x-troff-ms",
	     ".vcs", "text/x-vCalendar",
	     ".vcf", "text/x-vCard",

	     // misc apps
	     ".csm","application/cu-seeme",
	     ".cu", "application/cu-seeme",
	     ".tsp","application/dsptype",
	     ".spl","application/futuresplash",
	     ".pgp","application/pgp-signature",
             ".wz", "application/x-Wingz",
	     ".dcr","application/x-director",
	     ".dir","application/x-director",
	     ".dxr","application/x-director",
             ".hdf","application/x-hdf",
             ".mif","application/x-mif",
	     ".nc" ,"application/x-netcdf",
	     ".cdf","application/x-netcdf",
	     ".pac","application/x-ns-proxy-autoconfig",
	     ".swf","application/x-shockwave-flash",
	     ".swfl","application/x-shockwave-flash",
	     ".ustar","application/x-ustar",
	     ".src","application/x-wais-source",
	     ".torrent","application/x-bittorrent",

	     // msdos
	     ".com","application/x-msdos-program",
	     ".exe","application/x-msdos-program",
	     ".bat","application/x-msdos-program",

	     //microsoft apps
	     ".xls","application/excel",
	     ".dot","application/msword",
             ".ppt","application/powerpoint",

	     //binary files
	     ".bin","application/octet-stream",

           };
}
